File: ios/Runner/Api/SongHostApiImpl.swift
//
//  SongHostApiImpl.swift
//  Runner
//
//  Created by tk on 2024/01/25.
//

class SongHostApiImpl: SongHostApi {
    private let albumRepository = AlbumRepository()
    private let songRepository = SongRepository()
    
    func getAllSongs() throws -> [SongDTO] {
        let songs = songRepository.findAll()
        
        return songs.map { $0.toPigeon() }
    }
    
    func getSongsByArtistId(artistId: String) throws -> [SongDTO] {
        let albums = albumRepository.findByArtistId(artistId: ArtistId(id: artistId))
        let albumIds = albums.map { $0.id }
        let songs = songRepository.findByAlbumIds(albumIds: albumIds)

        return songs.map { $0.toPigeon() }
    }
    
    func getSongsByAlbumId(albumId: String) throws -> [SongDTO] {
        let songs = songRepository.findByAlbumId(albumId: AlbumId(id: albumId))
        
        return songs.map { $0.toPigeon() }
    }
    
    func getSongsByIds(ids: [String]) throws -> [SongDTO] {
        let songIds = ids.map { SongId(id: $0) }
        let songs = songRepository.findBySongIds(songIds: songIds)
        
        return songs.map { $0.toPigeon() }
    }
    
    func getSongById(id: String) throws -> SongDTO? {
        guard let song = songRepository.findBySongId(songId: SongId(id: id)) else {
            return nil
        }
        
        return song.toPigeon()
    }
}

File: android/app/src/main/kotlin/dev/tcode/thinmpf/api/SongHostApiImpl.kt
package dev.tcode.thinmpf.api

import android.content.Context
import dev.tcode.thinmpf.model.valueObject.AlbumId
import dev.tcode.thinmpf.model.valueObject.ArtistId
import dev.tcode.thinmpf.repository.SongRepository
import dev.tcode.thinmpf.extension.toPigeon
import dev.tcode.thinmpf.model.valueObject.SongId
import dev.tcode.thinmpf.pigeon_output.SongDTO
import dev.tcode.thinmpf.pigeon_output.SongHostApi

class SongHostApiImpl(context: Context) : SongHostApi {
    private val repository = SongRepository(context)

    override fun getAllSongs(): List<SongDTO> {
        val songs = repository.findAll()

        return songs.map { it.toPigeon() }
    }

    override fun getSongsByArtistId(artistId: String): List<SongDTO> {
        val songs = repository.findByArtistId(ArtistId(artistId))

        return songs.map { it.toPigeon() }
    }

    override fun getSongsByIds(ids: List<String>): List<SongDTO> {
        val songIds = ids.map { SongId(it) }
        val songs = repository.findByIds(songIds)

        return songs.map { it.toPigeon() }
    }

    override fun getSongsByAlbumId(albumId: String): List<SongDTO> {
        val songs = repository.findByAlbumId(AlbumId(albumId))

        return songs.map { it.toPigeon() }
    }

    override fun getSongById(id: String): SongDTO? {
        val songId = SongId(id)
        val song = repository.findById(songId) ?: return null

        return song.toPigeon()
    }
}